stl/inc/syncstream

```diff
@@ -0,0 +1,312 @@
+}
+void _Delete_mutex_for_instance_or_decrement(void* _Ptr) noexcept {
+    scoped_lock _Guard(_Get_map_mutex());
+    auto& _Mutex_map         = _Get_mutex_map();
+    auto _Instance_mutex_itr = _Mutex_map.find(_Ptr);
+    if (_Instance_mutex_itr == _Mutex_map.end())
+        return;
+    _Instance_mutex_itr->second._Ref_count--;
+    if (_Instance_mutex_itr->second._Ref_count == 0) {
+        _Mutex_map.erase(_Ptr);
+    }
+}
+
+// CLASS TEMPLATE basic_syncbuf
+template <class _Elem, class _Traits, class _Alloc>
+class basic_syncbuf : public basic_streambuf<_Elem, _Traits> {
+public:
+    using int_type       = typename _Traits::int_type;
+    using pos_type       = typename _Traits::pos_type;
+    using off_type       = typename _Traits::off_type;
+    using allocator_type = _Alloc;
+    using streambuf_type = basic_streambuf<_Elem, _Traits>;
+
+    using pointer   = typename allocator_traits<_Alloc>::pointer;
+    using size_type = typename allocator_traits<_Alloc>::size_type;
+
+    basic_syncbuf() : basic_syncbuf(nullptr){};
+
+    explicit basic_syncbuf(streambuf_type* _StrBuf) : basic_syncbuf(_StrBuf, _Alloc{}) {}
+
+    basic_syncbuf(streambuf_type* _StrBuf, const _Alloc& _Al) : _Wrapped(_StrBuf), _Al(_Al) {
+        if (_Wrapped != nullptr) {
+            _Add_mutex_for_instance_or_increment(static_cast<void*>(_Wrapped));
+        }
+        _Init();
+    }
+
+    basic_syncbuf(basic_syncbuf&& _Right) {
+        _Al = _STD move(_Right._Al);
+        _Assign_rv(_STD move(_Right));
+    }
+
+    ~basic_syncbuf() {
+        _Emit();
+        _Tidy();
+    }
+
+    basic_syncbuf& operator=(basic_syncbuf&& _Right) {
+        _Emit();
+        _Move_assign(_STD move(_Right), _Choose_pocma<_Alloc>{});
+        return *this;
+    }
+
+    void swap(basic_syncbuf& _Right) {
+        if (this != _STD addressof(_Right)) {
+            _Pocs(_Getal(), _Right._Getal());
+            _Swap_except_al(_Right);
+        }
+    }
+
+    bool emit() {
+        if (_Wrapped == nullptr) {
+            return false;
+        }
+
+        bool _Result                 = true;
+        const size_type _Data_size   = _Get_data_size();
+        const pointer _Begin_seq_ptr = pointer_traits<pointer>::pointer_to(*streambuf_type::pbase());
+        if (_Data_size > 0 || _Record_sync) {
+            scoped_lock _Guard(_Get_mutex_for_instance(static_cast<void*>(_Wrapped)));
+
+            if (_Data_size > 0 && _Data_size != _Wrapped->sputn(_Begin_seq_ptr, _Data_size)) {
+                _Result = false;
+            }
+
+            if (_Record_sync) {
+                if (_Wrapped->pubsync() == -1) {
+                    _Result = false;
+                }
+            }
+        }
+        _Record_sync = false;
+        streambuf_type::setp(
+            _Begin_seq_ptr, pointer_traits<pointer>::pointer_to(*streambuf_type::epptr())); // reset written data
+        return _Result;
+    }
+
+    streambuf_type* get_wrapped() const noexcept {

```

Comment:
General comment: For every non-`void` function (especially public, but also `_Ugly_internal`), consider whether it meets our criteria for being marked `_NODISCARD`; we are trying to do this for all new code.

The most common criterion for marking something as nodiscard is when a function is a "pure observer" - with no side effects, the only reason to call a pure observer is to get its return value, so it would make no sense to discard it. (Less common criteria are "acquires a raw resource like `allocate()`" or "is almost always an error if discarded like `std::remove_if()`".) If a function has side effects and there's even a 10% chance that it can be reasonably called without using its return value, then we don't mark it - avoiding false positive warnings is extremely important here.

We are permitted to mark things as nodiscard beyond the occurrences of `[[nodiscard]]` depicted in the Standard.